core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -333,46 +340,97 @@ private static void logIfContextNarrowedTimeout(
+      @Override
+      public void run() {
+        Status status = buildDeadlineExceededStatusWithRemainingNanos(remainingNanos);
+        delayedCancelOnDeadlineExceeded(status, observer);
+      }
     }
 
-    @Override
-    public void run() {
-      InsightBuilder insight = new InsightBuilder();
-      stream.appendTimeoutInsight(insight);
-      // DelayedStream.cancel() is safe to call from a thread that is different from where the
-      // stream is created.
-      long seconds = Math.abs(remainingNanos) / TimeUnit.SECONDS.toNanos(1);
-      long nanos = Math.abs(remainingNanos) % TimeUnit.SECONDS.toNanos(1);
-
-      StringBuilder buf = new StringBuilder();
-      buf.append("deadline exceeded after ");
-      if (remainingNanos < 0) {
-        buf.append('-');
+    return deadlineCancellationExecutor.schedule(
+        new LogExceptionRunnable(new DeadlineExceededNotifyApplicationTimer()),
+        remainingNanos,
+        TimeUnit.NANOSECONDS);
+  }
+
+  private Status buildDeadlineExceededStatusWithRemainingNanos(long remainingNanos) {
+    final InsightBuilder insight = new InsightBuilder();
+    stream.appendTimeoutInsight(insight);
+
+    long seconds = Math.abs(remainingNanos) / TimeUnit.SECONDS.toNanos(1);
+    long nanos = Math.abs(remainingNanos) % TimeUnit.SECONDS.toNanos(1);
+
+    StringBuilder buf = new StringBuilder();
+    buf.append("deadline exceeded after ");
+    if (remainingNanos < 0) {
+      buf.append('-');
+    }
+    buf.append(seconds);
+    buf.append(String.format(".%09d", nanos));
+    buf.append("s. ");
+    buf.append(insight);
+
+    return DEADLINE_EXCEEDED.augmentDescription(buf.toString());
+  }
+
+  private void delayedCancelOnDeadlineExceeded(Status status, Listener<RespT> observer) {
+    deadlineCancellationSendToServerFuture = startDeadlineSendCancelToServerTimer(status);
+    executeCloseObserverInContext(observer, status);
+  }
+
+  private void executeCloseObserverInContext(final Listener<RespT> observer, final Status status) {
+    class CloseInContext extends ContextRunnable {
+      CloseInContext() {
+        super(context);
+      }
+
+      @Override
+      public void runInContext() {
+        closeObserver(observer, status, new Metadata());
       }
-      buf.append(seconds);
-      buf.append(String.format(".%09d", nanos));
-      buf.append("s. ");
-      buf.append(insight);
-      stream.cancel(DEADLINE_EXCEEDED.augmentDescription(buf.toString()));
     }
+
+    callExecutor.execute(new CloseInContext());
   }
 
-  private ScheduledFuture<?> startDeadlineTimer(Deadline deadline) {
-    long remainingNanos = deadline.timeRemaining(TimeUnit.NANOSECONDS);
+  private void closeObserver(Listener<RespT> observer, Status status, Metadata trailers) {
+    if (!observerClosed) {
+      observerClosed = true;
+      observer.onClose(status, trailers);
+    }
+  }
+
+  private ScheduledFuture<?> startDeadlineSendCancelToServerTimer(final Status status) {
```

Comment:
nit: now that there is only one caller of this method, it would be more readable to merge it into its caller.